target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+
+	unsigned int cmu_ncode_in;
+	unsigned int cmu_divn3_sel;
+	unsigned int cmu_big_kvco;
+	unsigned int cmu_sel_prediv = 0x0;
+	unsigned int cmu_sel_div4 = 0x0;
+	unsigned int cmu_divn2 = 0x4;
+	unsigned int cmu_divn2_selb = 0x1;
+	unsigned int cmu_bypass_pi = 0x1;
+	unsigned int cmu_en_ssc = 0x0;
+	unsigned int cmu_fcode_in = 0x000;
+	unsigned int cmu_step_in = 0x145;
+	unsigned int cmu_tbase_in = 0x30E;
+	unsigned int cmu_ssc_order = 0x0;
+	unsigned int cmu_time2_rst_width = 0x1;
+	unsigned int cmu_time0_ck = 0x3;
+	unsigned int cmu_clkrdy = 0x1;
+	unsigned int cmu_lpf_rs = 0x2;
+	unsigned int cmu_en_center_in = 0x0;
+	unsigned int cmu_en_wd = 0x0;
+	unsigned int cmu_pi_i_sel = 0x3;
+	unsigned int cmu_sel_cp_i = 0x3;
+	unsigned int cmu_sel_cco = 0x1;
+	unsigned int cmu_ldo_sel = 0x6;
+	unsigned int cmu_lpf_cp = 0x1;
+	unsigned int cmu_cp_new_en = 0x1;
+	unsigned int cmu_ldo_en = 0x1;
+	unsigned int cmu_vc_dly = 0x0;
+	unsigned int cmu_en_ckoobs = 0x0;
+	unsigned int test_en = 0x0;
+
+	if (rate == 500000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
+		cmu_ncode_in = 0x74;
+		cmu_divn3_sel = 0x1;
+		cmu_big_kvco = 0x1;
+	} else if (rate == 300000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
+		cmu_ncode_in = 0x5C;
+		cmu_divn3_sel = 0x2;
+		cmu_big_kvco = 0x0;
+	} else {
+		pr_err("not allowed set clock %d to %luMHz\n", pll->id, rate / 1000000);
+		return 0;
+	}
+
+	ctrl0 = (cmu_fcode_in<<20)|(cmu_divn2<<12)|(cmu_ncode_in<<4)|
+		(cmu_bypass_pi<<3)|(cmu_sel_div4<<2)|(cmu_sel_prediv<<0);
+	ctrl1 = (cmu_divn3_sel<<27)|(cmu_divn2_selb<<26)|(cmu_en_ssc<<25)|
+		(cmu_step_in<<12)|(cmu_tbase_in<<0);
+	misc = (cmu_ssc_order<<31)|(cmu_time2_rst_width<<29)|(cmu_time0_ck<<26)|
+		(cmu_clkrdy<<24)|(cmu_big_kvco<<23)|(cmu_lpf_rs<<20)|
+		(cmu_en_center_in<<19)|(cmu_en_wd<<18)|(cmu_pi_i_sel<<14)|
+		(cmu_sel_cp_i<<10)|(cmu_sel_cco<<9)|(cmu_ldo_sel<<6)|
+		(cmu_lpf_cp<<5)|(cmu_cp_new_en<<4)|(cmu_ldo_en<<3)|
+		(cmu_vc_dly<<2)|(cmu_en_ckoobs<<1)|(test_en<<0);
+
+	spin_lock_irqsave(&pll->lock, flags);
+
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~(1 << 12), RTL838X_PLL_GLB_CTRL);
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~1, RTL838X_PLL_GLB_CTRL);
+
+	sw_w32(misc, RTL838X_PLL_CPU_MISC_CTRL);
+	sw_w32(ctrl0, RTL838X_PLL_CPU_CTRL0);
+	sw_w32(ctrl1, RTL838X_PLL_CPU_CTRL1);
+	
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | 1, RTL838X_PLL_GLB_CTRL);
+	while(!(sw_r32(RTL838X_PLL_GLB_CTRL) & (1 << 8)));
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | (1 << 12), RTL838X_PLL_GLB_CTRL);
+
+	spin_unlock_irqrestore(&pll->lock, flags);
+
+	return 0;
+}
+
+static long rtl838x_pll_round_rate(struct clk_hw *hw, unsigned long rate, unsigned long *parent_rate)
+{
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	if (pll->id == CLK_CPU) 
+		return (rate < 450000000) ? 300000000 : 500000000;
+	else if (pll->id == CLK_MEM) 
+		return (rate < 246875000) ? 193750000 : 300000000;
+	else
+		return 200000000;
+}
+
+static int rtl838x_pll_enable(struct clk_hw *hw)
+{
+	return 0;
```

Comment:
probably nice to add a comment that this is currently unknown or not possible? e.g. why do we only have stubs.